src/coreclr/gc/gc.cpp

```diff
@@ -8389,29 +8404,260 @@ void destroy_card_table (uint32_t* c_table)
     dprintf (2, ("Table Virtual Free : %Ix", (size_t)&card_table_refcount(c_table)));
 }
 
+void gc_heap::get_card_table_element_sizes (uint8_t* start, uint8_t* end, size_t sizes[total_bookkeeping_elements])
+{
+    memset (sizes, 0, sizeof(size_t) * total_bookkeeping_elements);
+    sizes[card_table_element] = size_card_of (start, end);
+    sizes[brick_table_element] = size_brick_of (start, end);
+#ifdef CARD_BUNDLE
+    if (can_use_write_watch_for_card_table())
+    {
+        sizes[card_bundle_table_element] = size_card_bundle_of (start, end);
+    }
+#endif //CARD_BUNDLE
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    if (gc_can_use_concurrent)
+    {
+        sizes[software_write_watch_table_element] = SoftwareWriteWatch::GetTableByteSize(start, end);
+    }
+#endif // FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    sizes[seg_mapping_table_element] = size_seg_mapping_table_of (start, end);
+#ifdef BACKGROUND_GC
+    if (gc_can_use_concurrent)
+    {
+        sizes[mark_array_element] = size_mark_array_of (start, end);
+    } 
+#endif //BACKGROUND_GC
+}
+
+void gc_heap::get_card_table_element_layout (uint8_t* start, uint8_t* end, size_t layout[total_bookkeeping_elements + 1])
+{
+    size_t sizes[total_bookkeeping_elements];
+    get_card_table_element_sizes(start, end, sizes);
+
+    size_t alignment[total_bookkeeping_elements + 1];
+    alignment[card_table_element] = 1;
+    alignment[brick_table_element] = 2;
+#ifdef CARD_BUNDLE
+    alignment[card_bundle_table_element] = 1;
+#endif //CARD_BUNDLE
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    alignment[software_write_watch_table_element] = sizeof(size_t);
+#endif //FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+    alignment[seg_mapping_table_element] = sizeof (uint8_t*);
+    // In order to avoid a dependency between commit_mark_array_by_range and this logic, it is easier to make sure
+    // pages for mark array never overlaps with pages in the seg mapping table. That way commit_mark_array_by_range
+    // will never commit a page that is already committed here for the seg mapping table.
+    alignment[mark_array_element] = OS_PAGE_SIZE;
+    // commit_mark_array_by_range extends the end pointer of the commit to the next page boundary, we better make sure it
+    // is reserved
+    alignment[total_bookkeeping_elements] = OS_PAGE_SIZE;
+
+    layout[card_table_element] = ALIGN_UP(sizeof(card_table_info), alignment[card_table_element]);
+    for (int element = brick_table_element; element <= total_bookkeeping_elements; element++)
+    {
+        layout[element] = layout[element - 1] + sizes[element - 1];
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+        if (element != software_write_watch_table_element || gc_can_use_concurrent)
+#endif //FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
```

Comment:
it's weird that this is only done for WW, but there are other elements in the same boat (like #ifdef BGC also checks for gc_can_use_concurrent) why is only WW special cased?